scipy/sparse/_coo.py

```diff
@@ -722,16 +810,463 @@ def _matmul_multivector(self, other):
+                              if ax not in axes_self)
+        og_shape_other = tuple(other.shape[ax] for ax in range(ndim_other)
+                               if ax not in axes_other)
+        
+        # Create 2D coords arrays
+        ravel_coords_shape_self = (math.prod(og_shape_self),
+                                math.prod([self.shape[ax] for ax in axes_self]))
+        ravel_coords_shape_other = (math.prod([other.shape[ax] for ax in axes_other]),
+                                    math.prod(og_shape_other))
+
+        self_2d_coords = (self_non_red_coords, self_reduced_coords)
+        other_2d_coords = (other_reduced_coords, other_non_red_coords)
+
+        self_2d = coo_array((self.data, self_2d_coords), ravel_coords_shape_self)
+        other_2d = coo_array((other.data, other_2d_coords), ravel_coords_shape_other)
+
+        # Perform matrix multiplication (routed via 2-D CSR)
+        prod = (self_2d @ other_2d).tocoo()
+
+        # Combine the shapes of the non-contracted axes
+        combined_shape = og_shape_self + og_shape_other
+
+        # Unravel the 2D coordinates to get multi-dimensional coordinates
+        if og_shape_self:
+            unraveled_coords_self = np.array(np.unravel_index(prod.row, og_shape_self))
+        if og_shape_other:
+            unraveled_coords_other = np.array(np.unravel_index(prod.col,
+                                                               og_shape_other))
+
+        if og_shape_self and og_shape_other:
+            coords = np.concatenate((unraveled_coords_self, unraveled_coords_other))
+        elif og_shape_self:
+            coords = unraveled_coords_self
+        elif og_shape_other:
+            coords = unraveled_coords_other
+        else:
+            coords = np.array([])
+
+        if coords.shape == (0,):  # if result is scalar
+            return prod.data[0]
+
+        # Construct the resulting COO array with combined coordinates and shape
+        prod_arr = coo_array((prod.data, tuple(coords)), shape=combined_shape)
+
+        return prod_arr
+
+
+    def _dense_tensordot(self, other, axes_self, axes_other):
+        ndim_self = len(self.shape)
+        ndim_other = len(other.shape)
+
+        non_reduced_axes_self = [ax for ax in range(ndim_self) if ax not in axes_self]
+        reduced_shape_self = [self.shape[s] for s in axes_self]
+        non_reduced_shape_self = [self.shape[s] for s in non_reduced_axes_self]
+
+        non_reduced_axes_other = [ax for ax in range(ndim_other)
+                                  if ax not in axes_other]
+        reduced_shape_other = [other.shape[s] for s in axes_other]
+        non_reduced_shape_other = [other.shape[s] for s in non_reduced_axes_other]
+
+        permute_self = non_reduced_axes_self + axes_self
+        permute_other = non_reduced_axes_other[:-1] + axes_other \
+            + non_reduced_axes_other[-1:]
+        self = np.transpose(self, permute_self)
+        other = np.transpose(other, permute_other)
```

Comment:
Does `np.transpose` work on a `scipy.sparse`?  Don't we have to reorder the `coords` instead?

Reply:
Corrected this now!